pilot/pkg/bootstrap/certcontroller.go

```diff
@@ -171,46 +163,119 @@ func (s *Server) initDNSCerts(hostname, customHost, namespace string) error {
-	// use to start grpc and webhooks is based on files. This is a memory-mounted dir.
-	if err := os.MkdirAll(dnsCertDir, 0o700); err != nil {
-		return err
-	}
-	err = ioutil.WriteFile(dnsKeyFile, keyPEM, 0o600)
-	if err != nil {
-		return err
-	}
-	err = ioutil.WriteFile(dnsCertFile, certChain, 0o600)
-	if err != nil {
-		return err
+// initCertificateWatches sets up watches for the dns certs.
+func (s *Server) initCertificateWatches(tlsOptions TLSOptions) error {
+	if hasCustomTLSCerts(tlsOptions) {
+		if err := s.keyCertBundleWatcher.SetFromFilesAndNotify(tlsOptions.KeyFile, tlsOptions.CertFile, tlsOptions.CaCertFile); err != nil {
+			return fmt.Errorf("set keyCertBundle failed: %v", err)
+		}
+		// TODO: Setup watcher for root and restart server if it changes.
+		for _, file := range []string{tlsOptions.CertFile, tlsOptions.KeyFile} {
+			log.Infof("adding watcher for certificate %s", file)
+			if err := s.fileWatcher.Add(file); err != nil {
+				return fmt.Errorf("could not watch %v: %v", file, err)
+			}
+		}
+		s.addStartFunc(func(stop <-chan struct{}) error {
+			go func() {
+				var keyCertTimerC <-chan time.Time
+				for {
+					select {
+					case <-keyCertTimerC:
+						keyCertTimerC = nil
+						if err := s.keyCertBundleWatcher.SetFromFilesAndNotify(tlsOptions.KeyFile, tlsOptions.CertFile, tlsOptions.CaCertFile); err != nil {
+							log.Errorf("Setting keyCertBundle failed: %v", err)
+						}
+					case <-s.fileWatcher.Events(tlsOptions.CertFile):
+						if keyCertTimerC == nil {
+							keyCertTimerC = time.After(watchDebounceDelay)
+						}
+					case <-s.fileWatcher.Events(tlsOptions.KeyFile):
+						if keyCertTimerC == nil {
+							keyCertTimerC = time.After(watchDebounceDelay)
+						}
+					case err := <-s.fileWatcher.Errors(tlsOptions.CertFile):
+						log.Errorf("error watching %v: %v", tlsOptions.CertFile, err)
+					case err := <-s.fileWatcher.Errors(tlsOptions.KeyFile):
+						log.Errorf("error watching %v: %v", tlsOptions.KeyFile, err)
+					case <-stop:
+						return
+					}
+				}
+			}()
+			return nil
+		})
 	}
-	log.Info("DNS certificates created in ", dnsCertDir)
+
+	notifyCh := make(chan struct{})
+	watchCh := s.keyCertBundleWatcher.AddWatcher()
+	go s.reloadCert(watchCh, notifyCh)
+	// wait for istiod cert first time loaded
+	<-notifyCh
 	return nil
 }
+
+func (s *Server) reloadCert(watchCh <-chan keycertbundle.KeyCertBundle, notifyCh chan struct{}) {
+	once := sync.Once{}
+	for {
+		keyCertBundle := <-watchCh
+		keyPair, err := tls.X509KeyPair(keyCertBundle.CertPem, keyCertBundle.KeyPem)
+		if err != nil {
+			log.Errorf("istiod loading x509 key pairs failed: %v", err)
+			once.Do(func() {
+				close(notifyCh) // This is for ut
+			})
+			continue
+		}
+		s.certMu.Lock()
+		s.istiodCert = &keyPair
+		s.certMu.Unlock()
+		once.Do(func() {
+			close(notifyCh)
+		})
+		var cnum int
+		log.Info("Istiod certificates are reloaded")
+		for _, c := range keyPair.Certificate {
```

Comment:
Shouldn't this check come before we load Istiod?

Reply:
Yes, i think so. BTW, it is also from legacy code. I just moved the location from `pilot/pkg/bootstrap/server.go` here

Reply:
done